File: conv3d.py
import torch
import torch.nn.functional as F
import numpy as np

# Function to write NumPy array to a binary file
def write_binary_file(file_path, array):
    with open(file_path, 'wb') as f:
        array.tofile(f)

INPUT_DEPTH = 3
INPUT_HEIGHT = 128
INPUT_WIDTH = 64
KERNEL_DEPTH = 3
KERNEL_HEIGHT = 3
KERNEL_WIDTH = 3

# Initialize input 3D matrix
temp = 0
input_matrix = []
for d in range(INPUT_DEPTH):
    depth_slice = []
    for i in range(INPUT_HEIGHT):
        row = []
        for j in range(INPUT_WIDTH):
            row.append(temp)
            temp += 1
        depth_slice.append(row)
    input_matrix.append(depth_slice)

# Initialize kernel 3D matrix
temp = 0
kernel_matrix = []
for d in range(KERNEL_DEPTH):
    depth_slice = []
    for i in range(KERNEL_HEIGHT):
        row = []
        for j in range(KERNEL_WIDTH):
            row.append(temp)
            temp += 1
        depth_slice.append(row)
    kernel_matrix.append(depth_slice)

# Convert lists to numpy arrays and then to PyTorch tensors
input_tensor = torch.tensor(input_matrix, dtype=torch.int).unsqueeze(0).unsqueeze(0)  # Shape: [1, 1, INPUT_DEPTH, INPUT_HEIGHT, INPUT_WIDTH]
kernel_tensor = torch.tensor(kernel_matrix, dtype=torch.int).unsqueeze(0).unsqueeze(0)  # Shape: [1, 1, KERNEL_DEPTH, KERNEL_HEIGHT, KERNEL_WIDTH]

# Perform 3D convolution using PyTorch
output_tensor = F.conv3d(input_tensor, kernel_tensor, stride=1, padding=0)

# Convert output tensor to 1D array
output_np = output_tensor.numpy()  # Convert to NumPy array
output_1d = output_np.flatten()    # Flatten to 1D array

print("Input: ",input_tensor.shape)
print("Kernel: ",kernel_tensor.shape)
print("Output: ",output_tensor.shape)
# Write the 1D array to a binary file
write_binary_file('py_conv3d_output.bin', output_1d)

print("3D Convolution complete. Output written to 'py_conv3d_output.bin'.")

File: conv3d.cpp
#include <iostream>
#include <vector>
#include <fstream>
#define INPUT_HEIGHT 128
#define INPUT_WIDTH 64
#define KERNEL_HEIGHT 3
#define KERNEL_WIDTH 3
#define CHANNEL 3

using namespace std;

void write_to_binary(const string &filename, const vector<int> &data)
{
    ofstream file(filename, ios::binary);
    file.write(reinterpret_cast<const char *>(data.data()), data.size() * sizeof(int));
    file.close();
}

int main()
{
    vector<vector<vector<int>>> input;
    vector<vector<vector<int>>> kernel;

    int temp = 0;
    for (int k = 0; k < CHANNEL; k++)
    {
        vector<vector<int>> col;
        for (int i = 0; i < INPUT_HEIGHT; i++)
        {
            vector<int> row;
            for (int j = 0; j < INPUT_WIDTH; j++)
            {
                row.push_back(temp);
                temp++;
            }
            col.push_back(row);
        }
        input.push_back(col);
    }

    temp = 0;
    for (int k = 0; k < CHANNEL; k++)
    {
        vector<vector<int>> col;
        for (int i = 0; i < KERNEL_HEIGHT; i++)
        {
            vector<int> row;
            for (int j = 0; j < KERNEL_WIDTH; j++)
            {
                row.push_back(temp);
                temp++;
            }
            col.push_back(row);
        }
        kernel.push_back(col);
    }

    int stride = 1;
    int padding = 0;

    int output_height = ((INPUT_HEIGHT - KERNEL_HEIGHT + (2 * padding)) / stride) + 1;
    int output_width = ((INPUT_WIDTH - KERNEL_WIDTH + (2 * padding)) / stride) + 1;

    cout << "Output Height: " << output_height << endl;
    cout << "Output Width: " << output_width << endl;

    vector<vector<vector<int>>> output(1, vector<vector<int>>(output_height, vector<int>(output_width, 0)));

    for (int ih = 0; ih < output_height; ih++)
    {
        for (int iw = 0; iw < output_width; iw++)
        {
            int sum = 0;
            for (int kc = 0; kc < CHANNEL; kc++)
            {
                for (int kh = 0; kh < KERNEL_HEIGHT; kh++)
                {
                    for (int kw = 0; kw < KERNEL_WIDTH; kw++)
                    {
                        int ow = iw * stride + kw;
                        int oh = ih * stride + kh;
                        sum += input[kc][oh][ow] * kernel[kc][kh][kw];
                    }
                }
            }
            output[0][ih][iw] = sum;
        }
    }

    vector<int> output_1d;
    for (int j = 0; j < output_height; j++)
    {
        for (int k = 0; k < output_width; k++)
        {
            output_1d.push_back(output[0][j][k]);
        }
    }

    write_to_binary("cpp_conv3d_output.bin", output_1d);

    cout << "3D Convolution complete. Output written to 'cpp_conv3d_output.bin'." << endl;

    return 0;
}
